src/compiler/compiler.rs

```diff
@@ -275,124 +296,205 @@ impl Compiler {
-                                 }));
-                        }
-                        if !compiler_result.stdout.is_empty() {
-                            try!(entry.put_object("stdout", &mut io::Cursor::new(&compiler_result.stdout)));
-                        }
-                        if !compiler_result.stderr.is_empty() {
-                            try!(entry.put_object("stderr", &mut io::Cursor::new(&compiler_result.stderr)));
-                        }
-                        // Try to finish storing the newly-written cache
-                        // entry. We'll get the result back elsewhere.
-                        let out_file = out_file.into_owned();
-                        let future = storage.finish_put(&key, entry)
-                            .then(move |res| {
-                                match res {
-                                    Ok(_) => debug!("[{}]: Stored in cache successfully!", out_file),
-                                    Err(ref e) => debug!("[{}]: Cache write error: {:?}", out_file, e),
-                                }
-                                res.map(|duration| CacheWriteInfo {
-                                    object_file: out_file,
-                                    duration: duration,
-                                })
-                            });
-                        let future = Box::new(future);
-                        Ok((CompileResult::CacheMiss(miss_type, duration, future), compiler_result))
-                    } else {
-                        // Not cacheable
-                        debug!("[{}]: Compiled but not cacheable", out_file);
-                        Ok((CompileResult::NotCacheable, compiler_result))
-                    }
+                me.compile(&creator,
+                           preprocessor_result,
+                           parsed_args,
+                           &cwd,
+                           pool,
+                           outputs,
+                           storage,
+                           key,
+                           miss_type)
+            }))
+        }))
+    }
+
+    fn compile<T>(&self,
+                  creator: &T,
+                  preprocessor_result: process::Output,
+                  parsed_args: ParsedArguments,
+                  cwd: &str,
+                  pool: CpuPool,
+                  outputs: HashMap<String, PathBuf>,
+                  storage: Arc<Storage>,
+                  key: String,
+                  miss_type: MissType)
+                  -> Box<Future<Item=(CompileResult, process::Output),
+                                Error=io::Error>>
+        where T: CommandCreatorSync,
+    {
+        let process::Output { stdout, .. } = preprocessor_result;
+        let start = Instant::now();
+
+        let compile = self.kind.compile(creator, self, stdout, &parsed_args, cwd, &pool);
+        Box::new(compile.and_then(move |(cacheable, compiler_result)| {
```

Comment:
Ah this is just stylistic for me. I prefer to avoid rightward-drift and keep things under 80 cols, so I tend to make more local variables to avoid the drift and indentation. Both work equivalently though of course!